modin/pandas/dataframe.py

```diff
@@ -2451,13 +2452,22 @@ def __setitem__(self, key, value):
                         value = np.array(value)
                     if len(key) != value.shape[-1]:
                         raise ValueError("Columns must be same length as key")
-                item = broadcast_item(
-                    self,
-                    slice(None),
-                    key,
-                    value,
-                    need_columns_reindex=False,
-                )
+
+                # If we have a modin Series or DataFrame, let's not convert it to
+                # a numpy array if we can help it.
+                if isinstance(value, (DataFrame, Series)):
+                    item = _maybe_reindex_item(
+                        self, slice(None), key, value, need_columns_reindex=False
+                    )
+                else:
+                    item = broadcast_item(
```

Comment:
So `_maybe_reindex_item` gets called in either code path here (either directly here, or in `broadcast_item`)? Can we always call it directly here?

Reply:
Yeah actually this was dumb thinking on my part. I could just return early in `broadcast_item`.